### PR TITLE
feat: adds interactive lab embed

### PR DESCRIPTION
**Pull request by kendallstrautman**
## "Ready for Review" checklist

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

- [Preview link](https://dev-portal-git-ksinstruqt-embed-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-docker) 🔎
- [Asana task](https://app.asana.com/0/1201987349274776/1201992883045078) 🎟️

## What

This PR adds the interactive [instruqt embed](https://instruqt.com/) for the interactive lab tutorial feature. 

Screen Shot 2022-03-18 at 4 07 57 PM

## Why

Feature parity! Many learn tutorials [have interactive labs](https://learn.hashicorp.com/search?isInteractive=true&page=1) so this is a base requirement. 

## How

Most of the code is copied directly from Learn. I converted the components to typescript, but didn't really change anything within.

## Testing

- [Visit this tutorial](https://dev-portal-git-ksinstruqt-embed-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-)
- Hit the _show terminal_ button
- The lab embed should open, try launching the lab. It's the work of instruqt after that. 
- Compare with [functionality on Learn here](https://learn.hashicorp.com/tutorials/waypoint/get-started-docker)

## Anything else?

This only supports instruqt embeds and fully deprecates the katacoda usecase. In looking through the tutorials repo, that lab provider isn't used anymore, so we can officially rip it out with dev dot.

### CODE REVIEW

File: src/components/lab-embed/embed-element/index.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { MutableRefObject, useRef, useEffect } from 'react'
+import { useInstruqtEmbed } from '../providers/instruqt'
+import styles from './embed-element.module.css'
+import { HandsOnLabProviderOption } from 'lib/learn-client/types'
+
+export default function EmbedElement({
+  provider,
+}: {
+  provider: HandsOnLabProviderOption
```
Comment (kendallstrautman) [inline src/components/lab-embed/embed-element/index.tsx:9]: Should be instruqt only

File: src/components/lab-embed/embed-element/index.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { MutableRefObject, useRef, useEffect } from 'react'
```
Comment (kendallstrautman) [inline src/components/lab-embed/embed-element/index.tsx:1]: See the [lab embed in Learn](https://github.com/hashicorp/learn/tree/master/components/lab-embed), copied from there.

File: src/components/lab-embed/resizable/components/img/resize_bar.svg [hunk 3]
```diff
@@ -0,0 +1 @@
+<svg width="120" height="14" fill="none" xmlns="http://www.w3.org/2000/svg"><rect width="120" height="4" rx="2" fill="var(--gray-5, #dbdbdc)"/><rect y="10" width="120" height="4" rx="2" fill="var(--gray-5, #dbdbdc)"/></svg>
```
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/lab-embed/resizable/components/img/resize_bar.svg:1]: Screen Shot 2022-03-18 at 4 07 57 PM

File: src/components/lab-embed/resizable/components/resizer.module.css [hunk 4]
```diff
@@ -0,0 +1,35 @@
+.resizer {
+  width: 100%;
+  height: 40px;
+  position: absolute;
+  cursor: row-resize;
+  top: -5px;
+  z-index: 10;
+  background: var(--token-color-palette-neutral-50);
```
Comment (kendallstrautman) [inline src/components/lab-embed/resizable/components/resizer.module.css:8]: I updated these colors to use design token values. Just sort of eye-balled it.

**Comment by kendallstrautman** [comment]
I just rebased this back onto main to get the vault routes and now the deploy is failing because of a packer docs thing. Looking into it.

File: src/views/tutorial-view/index.tsx [hunk 5]
```diff
@@ -74,48 +77,60 @@ export default function TutorialView({
...
+            tutorialMeta={{
+              readTime,
+              products: productsUsed.map((p) => p.product.slug),
+              isBeta: getIsBeta(productsUsed),
+              edition,
+              hasVideo: Boolean(video),
+              isInteractive: Boolean(handsOnLab),
+            }}
+          />
+          <InteractiveLabToggle showButton={Boolean(handsOnLab?.id)} />
```
Comment (ashleemboyer) [inline src/views/tutorial-view/index.tsx:124]: Thoughts on removing the `showButton` prop and conditionally rendering this component instead?

```tsx
{isInteractive && <InteractiveLabToggle />}
```

Also seeing that the `isInteractive` logic is defined in two different ways, we probably want just one:

- `Boolean(handsOnLab)` (in the `tutorialMeta` object just above this)
- `Boolean(handsOnLab?.id)`

File: src/views/tutorial-view/index.tsx [hunk 6]
```diff
@@ -74,48 +77,60 @@ export default function TutorialView({
     video,
     collectionCtx,
   } = tutorial
+
+  const InteractiveLabWrapper = handsOnLab?.id ? InstruqtProvider : Fragment
```
Comment (ashleemboyer) [inline src/views/tutorial-view/index.tsx:81]: Relevant to previous comment, I think you could define a `isInteractive` boolean variable above this and use it on this line and down below in the markup where my other comment is.

**Review by ashleemboyer — CHANGES_REQUESTED** [review]
This is a really great start! ✨ Love learning about how this interactive frame works. Have a couple of areas of feedback and a couple of comments just thinking out loud.

File: src/components/lab-embed/resizable/components/img/resize_bar.svg (hunk 3 above)
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/lab-embed/resizable/components/img/resize_bar.svg:1]: Sounds good! I'll adjust the color

File: src/components/lab-embed/resizable/components/resizer.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+import ResizeBar from './img/resize_bar.svg?include'
+import s from './resizer.module.css'
+
+interface ResizerProps {
+  onMouseDown(e: React.MouseEvent): void
+  panelActive: boolean
+  style: CSS.Properties
+  onMouseUp?(e: React.MouseEvent): void
+  onClosePanel?: Dispatch<SetStateAction<boolean>>
+}
```
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/lab-embed/resizable/components/resizer.tsx:14]: Sure! Sounds good

File: src/components/lab-embed/resizable/index.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      )}
+      ref={resizableDiv}
+      style={{
+        height: `${height}px`,
+      }}
+      data-resizing={String(isResizing)}
+    >
+      <Resizer
+        onClosePanel={setPanelActive}
+        panelActive={panelActive}
```
[the remark this answers is not in the document]
Comment (kendallstrautman) [reply src/components/lab-embed/resizable/index.tsx:91]: Cool, I'll refactor

**Comment by kendallstrautman** [comment]
@ashleemboyer thank you 🙇 I think I addressed most of your feedback. lmk if you see anything else!

File: src/views/tutorial-view/index.tsx [hunk 9]
```diff
@@ -74,48 +77,60 @@ export default function TutorialView({
...
+            {name}
+          </Heading>
+          <Badges
+            tutorialMeta={{
+              readTime,
+              products: productsUsed.map((p) => p.product.slug),
+              isBeta: getIsBeta(productsUsed),
+              edition,
+              hasVideo: Boolean(video),
+              isInteractive: Boolean(handsOnLab),
```
Comment (ashleemboyer) [inline src/views/tutorial-view/index.tsx:121]:
```suggestion
              isInteractive,
```

**Review by ashleemboyer — APPROVED** [review]
One last suggestion, but approving so you can merge when ready! 🥇

File: src/views/tutorial-view/index.tsx (hunk 9 above)
Comment (kendallstrautman) [reply src/views/tutorial-view/index.tsx:121]: ty! 🙇

**Pull request merged**